libparsec/crates/client_connection/tests/proxy/mod.rs

```diff
@@ -0,0 +1,166 @@
+            Err(expect_err)
+        }
+    }
+}
+
+struct ProxyServer {
+    port: u16,
+    backend_addr: BackendAddr,
+    notify_disconnect: mpsc::Receiver<Event>,
+}
+
+impl ProxyServer {
+    async fn run(mut self, ready: oneshot::Sender<()>) -> io::Result<()> {
+        let url = self.backend_addr.to_http_url(None);
+        let server_addr = extract_url_domain_and_port(&url).ok_or_else(|| {
+            io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!("Cannot extract domain and port on url `{url}`"),
+            )
+        })?;
+        log::trace!("backend url: {url}, {server_addr:?}");
+        let listener = TcpListener::bind(("localhost", self.port)).await?;
+        ready
+            .send(())
+            .expect("The other side is waiting for this message");
+
+        loop {
+            let (mut client_socket, client_addr) = tokio::select! {
+                res = listener.accept() => {
+                    res
+                }
+                res = self.notify_disconnect.recv() => {
+                    match res {
+                        Some(Event::Disconnect) => {
+                            continue
+                        },
+                        Some(Event::Close) | None => {
+                            break
+                        },
+                    }
+                }
+            }?;
+            log::trace!("New client {}", client_addr);
+            let mut server_socket = match TcpStream::connect(server_addr).await {
+                Ok(v) => {
+                    log::trace!("Connected to backend");
+                    v
+                }
+                Err(e) => {
+                    log::error!("Failed to connect to the backend: {e}");
+                    drop(client_socket);
+                    continue;
+                }
+            };
+            tokio::select! {
+                copy_res = tokio::io::copy_bidirectional(&mut client_socket, &mut server_socket) => {
+                    log::warn!("Client / Server finished communicating ({copy_res:?})");
+                }
+                res = self.notify_disconnect.recv() => {
+                    match res {
+                        Some(Event::Disconnect) => {
+                            log::debug!("Disconnect client from proxy ...");
+                            drop(client_socket);
+                            drop(server_socket);
+                            continue;
+                        },
+                        Some(Event::Close) => {
+                            log::debug!("Closing proxy ...");
+                            drop(client_socket);
+                            drop(server_socket);
+                            break
+                        },
+                        None => {},
+                    }
+                }
+            }
+        }
+        log::info!("Proxy stopped");
+        Ok(())
+    }
```

Comment:
Can you space a bit ?

Reply:
Does the loop block need spaces or the entire fn block ?

Reply:
If you can space between log :thinking: